ENH: Add `subok` parameter to np.copy function (cf. #6509)

rossbar opened the pull request:
This is largely a re-submission of the original change proposed in #6509. Discussion was hosted in multiple forums including #3474, [the numpy mailing list circa 10-2015](https://mail.python.org/pipermail/numpy-discussion/2015-October/074021.html), and the [02-26-2020 NumPy Triage meeting](https://github.com/rossbar/archive/blob/triage_02-26-2020/triage_meetings/triage-2020-02-26.md).

This PR closes #3474 and #15570

rossbar committed:
Add subok parameter to np.copy function.

Re-introduces changes originally proposed in #6509 which was
closed due to very long review period.

rossbar committed:
Add See Also to docs highlighting preferred copy method

rossbar committed:
Fixed tests.

charris commented:
Needs a release note.

rossbar committed:
Added release note.

eric-wieser commented on doc/release/upcoming_changes/15685.new_feature.rst:
```diff
@@ -0,0 +1,7 @@
+``subok`` option for `np.copy`
+------------------------------
+A new kwarg, ``subok``, was added to `np.copy` to allow users to toggle the 
```
```suggestion
A new kwarg, ``subok``, was added to `numpy.copy` to allow users to toggle the 
```

Links don't work if they start with `np` instead of `numpy`.

rossbar committed:
np -> numpy in release note

Incorporate rst formatting feedback

mattip commented on doc/release/upcoming_changes/15685.new_feature.rst:
```diff
@@ -0,0 +1,7 @@
+behavior of `numpy.copy` with respect to array subclasses. The default value
+is ``False`` which is consistent with the behavior of `numpy.copy` for 
+previous numpy versions. To create a copy that preserves an array subclass with
+`numpy.copy`, call ``np.copy(arr, subok=True)``.
```
Maybe add "which can be preferably spelled `arr.copy()` ?

rossbar replied:
I added a sentence to the end of the release note to highlight the difference in the default behavior between `np.copy` and the `np.ndarray.copy` method. Let me know if that change was an improvement.

mattip commented on numpy/lib/function_base.py:
```diff
@@ -728,12 +728,21 @@ def copy(a, order='K'):
         arguments.)
+    subok : bool, optional
+        If True, then sub-classes will be passed-through, otherwise the
+        returned array will be forced to be a base-class array (default).
```
```suggestion
        returned array will be forced to be a base-class array (defaults to False).
```

rossbar replied:
I took the wording here straight out of the `np.array` docstring. If we change it here it should be changed everywhere else the parameter description appears verbatim, which (according to grep) is `array`, `broadcast_arrays` and `broadcast_to`.

I think your wording is indeed more clear - let me know if you think all the doc strings should be modified or if we should leave it as-is.

mattip replied:
I think we should keep the change local. Making our use of "default" more consistent across the code base is a worthy goal, but a bit beyond this PR.

rossbar committed:
Updated release note.

Added sentence to highlight difference bewteen np.copy and
np.ndarray.copy method.

rossbar committed:
Update numpy/lib/function_base.py

Update docstring "default" to be more explicit

Co-Authored-By: Matti Picus <matti.picus@gmail.com>

rossbar commented:
Okay, that should be everything - if the updated release note needs some re-wording let me know (or feel free to make the suggestion directly).
